### PR TITLE
Add modal to api errors

### LINKED ISSUE #21: Add state management library

### PR DESCRIPTION
**Pull request by erikskaar**
closes #21 #19

### CODE REVIEW

**Comment by erikskaar** [comment]
I would like to have the state get updated whenever the api wrapper throws an error, but Recoil doesn't allow updating state from outside the component tree unless you use third party add-ons. That's why it sets state directly in OntologyTable.tsx rather than directly from the API.

**Review by AndreasWintherMoen — CHANGES_REQUESTED** [review]
This doesn't really handle generic API errors though. Take a look at the ApiError class in the frontend's api folder. Whenever a non-200 status code is returned from the backend, an ApiError is thrown. Try just inserting some dimwitted code like `req.params.afh.asdf()` in the backend route to make it crash. This will also crash the frontend with the API error and stacktrace on the whole screen. To prevent this we need to add a try catch block in the frontend whenever we make an API request (such as line 52 in OntologyTable). The error that is caught in the catch block should then be dispatched to Recoil. 

In my opinion this would clutter the UI a bit too much though, so ideally we'd want all that in a separate function. Maybe have a folder with functions to handle API requests and dispatching possible errors, and this function is what we call in the UI component (Or maybe just do it in the API requests in the api folder?).  That way we don't need to deal with any error handling in the UI. However, it seems like Recoil has a limitation of only being accessible via UI components, i.e. we can't dispatch something to the state in a regular JS function. This seems like a pretty big disadvantage to me, and I think we should reconsider Redux..

Oh, and it would also be useful to have an errorMessage atom instead of just a boolean. And then we can display that message in the modal instead of just a default error message.

File: frontend/src/components/OntologyTable.tsx [hunk 1]
```diff
@@ -43,11 +45,17 @@ const renderOntology = (
 const OntologyTable: React.FC = () => {
   const [selectedNode, setSelectedNode] = useState(initialNode);
   const [ontologies, setOntologies] = useState([]);
+  const [, setApiError] = useRecoilState(apiErrorState);
```
Comment (AndreasWintherMoen) [inline frontend/src/components/OntologyTable.tsx:48]: When you don't need the value (just the setter function) you can use useSetRecoilState instead. There's also useRecoilValue when you only need the value and not the setter.

**Comment by AndreasWintherMoen** [comment]
There is a library called [recoil-outside](https://www.npmjs.com/package/recoil-outside) which supposedly lets you change the Recoil state outside of React components. It only has 24 weekly downloads though.

**Comment by erikskaar** [comment]
Yep, that is the one i was referring to. I will check it out to see if it's usable :)

**Comment by erikskaar** [comment]
I've updated it to use Redux.  `api.ts`  will now update the state, which `ErrorModal.tsx` is subscribed to, directly.

File: frontend/src/api/api.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+// eslint-disable-next-line import/no-cycle
```
Comment (AndreasWintherMoen) [inline frontend/src/api/api.ts:1]: We should consider adding this as an exception in the eslint config.

**Review by AndreasWintherMoen — APPROVED** [review]
Yea I think it's better like this with Redux. 

I've made a few changes. I think we should add store dispatch wrappers so that the rest of the app never has to import store and dispatch actions to it. Redux may be confusing for people who haven't used it and it's just unnecessarily verbose to write _store.dispatch_ everywhere. That's my opinion anyway, see if you agree. I've also changed the ErrorModal to use hooks.

**Comment by AndreasWintherMoen** [comment]
The frontend test fails because it expects reducer to be a function? But it is a function, right?

**Comment by erikskaar** [comment]
The tests are only example tests right? If they aren't needed we can just remove them.

**Comment by AndreasWintherMoen** [comment]
Well it tests that the app renders without crashing, which is kinda necessary. I think it might be related to the cycle imports.

**Comment by AndreasWintherMoen** [comment]
Ok I've fixed the error. It was indeed related to cycle imports. We should try to avoid that if possible. I've moved some stuff around and the types file is getting pretty clustered. We should try to fix that later (see #25).

**Comment by AndreasWintherMoen** [comment]
Btw try adding me as co author when you merge. I've never used it before but it looks like it tags both of us as contributors to the commit.

**Pull request merged**